### DISCUSSION TITLE
Why should width be explicitly mentioned in SIMD

### DISCUSSION THREAD

**Discussion by alfie-nsugh** [modular/modular#2330] [Q&A] [answered]
In this code 
```
from tensor import Tensor, TensorSpec, TensorShape
def main():
    alias type = DType.int64
    var spec = TensorSpec(type,5)
    var s = Tensor[type](spec)
    var t = Tensor[type](spec)
    var u = Tensor[type](spec)
    var slist = List(1,2,3,4,5)
    var tlist = List(2,3,5,7,11)
    for i in range(5):
        s[i] = slist[i]
        t[i] = tlist[i]
    u.store[5](0,s.load[width=5](0))
    print(u)
```
I see that the result is: `Tensor([[1, 2, 3, 4, 5]], dtype=int64, shape=5)` which is what I want.

But I get for 
```
def main():
    alias type = DType.int64
    var spec = TensorSpec(type,5)
    var s = Tensor[type](spec)
    var t = Tensor[type](spec)
    var u = Tensor[type](spec)
    var slist = List(1,2,3,4,5)
    var tlist = List(2,3,5,7,11)
    for i in range(5):
        s[i] = slist[i]
        t[i] = tlist[i]
    u.store[5](0,s.load[5](0))
    print(u)
```
This result
`Tensor([[1, 0, 0, 0, 0]], dtype=int64, shape=5)`

It took quite a while to figure that out. Is this a bug or does it make sense that we need to be this explicit with the width?

**Comment by rarebreed** [comment, marked as answer]
I would call that a bug.  

Mojo appears to be calling the overloaded `load` method of `load[len: Int, width: Int](self: Self, indices: StaticIntTuple[len]) -> SIMD[dtype, width]` which it should not.

According to the manual's rules:

> 1. Candidates with the minimal number of implicit conversions (in both arguments and parameters).
> 2. Candidates without variadic arguments.
> 3. Candidates without variadic parameters.
> 4. Candidates with the shortest parameter signature.
> 5. Non-@staticmethod candidates (over @staticmethod ones, if available).

The load method of `load[width: Int](self: Self, index: Int) -> SIMD[dtype, width]` should have been called because rule 1 should have been applied.

**Reply by rarebreed** [reply]
And actually, it looks like the `store` method is also calling the wrong overloaded method too.
